### ISSUE TITLE
Bug or feature - Confusing shapes of Tensor resulted from `tf.decode_csv`

### ISSUE THREAD

**Issue by BaiGang** [tensorflow/tensorflow#6554]
```
reader = tf.TextLineReader()
key, value = reader.read(filename_queue)
parsed = tf.decode_csv(value, [[0.0]] * 24)
label = parsed[0]
weight = parsed[1]
feature = tf.pack(parsed[2:])
label_tensor, weight_tensor, feature_tensor = tf.train.shuffle_batch(
      [label, weight, feature],
      batch_size=200,
      capacity=capacity,
      min_after_dequeue=min_after_dequeue,
      num_threads=num_threads)
print("Parsed tensors: %s, %s, %s" % )
```

The shapes of `label_tensor`, `weight_tensor` and `feature_tensor` are **`(200, )`**, **`(200, )`** and **`(200, 22)`** respectively.

If we put an extra comma at the end of line 4 and 5, like this:
```
label = parsed[0],
weight = parsed[1],
```
Then the shapes become **`(200, 1)`**, **`(200, 1)`** and **`(200, 26)`** respectively.

So an extra comma determines the shape of the resulted tensor. I am not quite familiar with Python and am new to TensorFlow, but is it supposed so?

**Comment by mrry** [comment]
The extra comma is significant: it turns a `tf.Tensor` object (e.g. `parsed[0]`) into a single-element `tuple` (e.g. `(parsed[0],)`).  When you pass them to APIs that expect a single `tf.Tensor`, TensorFlow automatically packs lists and tuples of `tf.Tensor` objects into a single `tf.Tensor` with an additional dimension. Therefore, since `parsed[0]` is a scalar (a 0-dimensional tensor), `parsed[0],` is treated as a single-element vector (a 1-dimensional tensor).

The `tf.train.shuffle_batch(..., batch_size=200, ...)` method accumulates a batch of 200 labels, weights, and features, by packing together 200 randomly sampled parsed inputs. Batching 200 scalars gives a shape of `(200,)`, whereas batching 200 single-element vectors gives a shape of `(200, 1)`.

Hope this helps. Note that in future, queries like this one are better handled on Stack Overflow, so that they're more easily searchable.

**Comment by BaiGang** [comment]
Thanks Derek for the detailed explanation. My question is resolved.
